rholang/src/main/scala/coop/rchain/rholang/interpreter/Env.scala

```diff
@@ -1,33 +1,41 @@
 package coop.rchain.rholang.interpreter
 
-import scala.collection.immutable.HashMap
+import scala.collection.mutable
 
-/* Env reifies the Env[A] = HashMap[Int,A] type alias.
+/* Env reifies the Env[A] = mutable.LinkedHashMap[Int,A] type alias.
    It extends HashMap with functions that manipulate
    the map based on level indices. */
 
 object Env {
 
-  def empty = HashMap.empty[Int, Par]
+  type Env[A] = mutable.LinkedHashMap[Int, A]
 
-  implicit class DeBruijn(env: HashMap[Int,Par]) {
+  def apply[A](a: A): Env[A] =
+    DeBruijn(mutable.LinkedHashMap.empty[Int, A]) + a
 
-    def get(level: Int): Option[Par] = env.get(level)
+  def apply[A](a: A, b: A, k: A*): Env[A] =
+    DeBruijn(mutable.LinkedHashMap.empty[Int, A]) + (a, b, k: _*)
 
-    def put(level: Int)(data: Par): HashMap[Int,Par] = env + (level -> data)
+  def apply[A](elems: (Int, A)*): Env[A] =
+    mutable.LinkedHashMap[Int, A](elems: _*)
 
-    /* Since indexing starts from 0, # of bound variables = level of
-       environment associated with the term.*/
-    def level: Int = env.keys.max + 1
+  implicit class DeBruijn[A](env: mutable.LinkedHashMap[Int, A]) {
 
-    def rename(j: Int): HashMap[Int,Par] = env map { case (k, data) => (k + j, data) }
+    def level: Int = if (env.isEmpty) 0 else env.last._1 + 1
 
-      /* In [env] allocate [k.size] new slots */
-    def alloc(k: List[Par]): HashMap[Int,Par] = (env /: k) { (_env, data) =>
-        _env.put(_env.level)(data)
-      }
+    def rename(j: Int): mutable.LinkedHashMap[Int, A] = env map {
+      case (k, data) => (k + j, data)
+    }
 
-    def merge(_env: HashMap[Int,Par]): HashMap[Int,Par] = env ++: _env.rename(env.level)
+    def +(a: A): mutable.LinkedHashMap[Int, A] = env += (env.level -> a)
```

Comment:
OK, then I'm really troubled by this line.
We should either expose it as +=, or we should delegate to +.

Reply:
Ah. I see now. Good point.

Reply:
I'll expose as `+=` since that's what the underlying operator is.